enterprise/frontend/src/embedding-sdk/components/public/InteractiveQuestion/components/Notebook.tsx

```diff
@@ -1,48 +1,48 @@
+import { useMemo } from "react";
+
 import { useInteractiveQuestionContext } from "embedding-sdk/components/public/InteractiveQuestion/context";
-import { useDispatch, useSelector } from "metabase/lib/redux";
-import {
-  runQuestionQuery,
-  updateQuestion,
-} from "metabase/query_builder/actions";
+import { useSelector } from "metabase/lib/redux";
 import { default as QBNotebook } from "metabase/query_builder/components/notebook/Notebook";
 import {
-  getIsDirty,
-  getIsResultDirty,
-  getIsRunnable,
+  isQuestionDirty,
+  isQuestionRunnable,
 } from "metabase/query_builder/selectors";
 import { getSetting } from "metabase/selectors/settings";
 import { ScrollArea } from "metabase/ui";
-import type Question from "metabase-lib/v1/Question";
 
 type NotebookProps = { onApply?: () => void };
 
 export const Notebook = ({ onApply = () => {} }: NotebookProps) => {
-  const { question } = useInteractiveQuestionContext();
+  const { question, originalQuestion, updateQuestion, runQuestion } =
+    useInteractiveQuestionContext();
+
+  const isDirty = useMemo(() => {
+    return isQuestionDirty(question, originalQuestion);
+  }, [question, originalQuestion]);
+
+  const isRunnable = useMemo(() => {
+    return isQuestionRunnable(question, isDirty);
+  }, [question, isDirty]);
 
-  const isDirty = useSelector(getIsDirty);
-  const isRunnable = useSelector(getIsRunnable);
-  const isResultDirty = useSelector(getIsResultDirty);
   const reportTimezone = useSelector(state =>
     getSetting(state, "report-timezone-long"),
   );
 
-  const dispatch = useDispatch();
-
   return (
     question && (
       <ScrollArea w="100%" h="100%">
         <QBNotebook
           question={question}
           isDirty={isDirty}
           isRunnable={isRunnable}
-          isResultDirty={Boolean(isResultDirty)}
```

Comment:
We do not have the necessary metadata for `isResultDirty` (e.g. `tableMetadata`), so we use `isDirty` for now. This should suffice for now, as `isResultDirty` determines whether to run the question's query or not. In this case, it would always re-run the question query when "Visualize" is clicked, which is an acceptable trade-off.

```ts
  // visualize switches the view to the question's visualization.
  async function visualize() {
    // Only cleanup the question if it's dirty, otherwise Metabase
    // will incorrectly display the Save button, even though there are no changes to save.
    if (isDirty) {
      cleanupQuestion();
    }
    // switch mode before running otherwise URL update may cause it to switch back to notebook mode
    await setQueryBuilderMode("view");
    if (isResultDirty) {
      await runQuestionQuery();
    }
  }

```